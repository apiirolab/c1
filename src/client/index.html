<!DOCTYPE html>
<html>

<head>
  <title></title>
  <meta name="description" content="Corporate process management for Western Sydney University!">
  <link rel="icon" type="image/png" href="/favico.png">
  <link href='https://cdnjs.cloudflare.com/ajax/libs/semantic-ui/2.3.2/semantic.min.css' rel='stylesheet' type='text/css'>
</head>

<body>
  <div id="root"></div>
  $bundles
</body>

</html>