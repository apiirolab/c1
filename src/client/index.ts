export const info = 'Client Info';
