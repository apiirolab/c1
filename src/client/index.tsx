import * as React from 'react';
import * as ReactDOM from 'react-dom';

import { ApolloProvider } from 'react-apollo';
import { Routes } from './config/router';
import { client } from './config/apollo';
// import * as serviceWorker from './registerServiceWorker';

ReactDOM.render(
  <ApolloProvider client={client}>
    <Routes />
  </ApolloProvider>,
  document.getElementById('root')
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
// serviceWorker.unregister();
