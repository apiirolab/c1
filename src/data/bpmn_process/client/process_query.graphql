query ProcessQuery($id: ID!) {
  process(id: $id) {
    actionCount
    description
    id
    model
    resources {
      name
      type
    }
    status
    version
  }
}
