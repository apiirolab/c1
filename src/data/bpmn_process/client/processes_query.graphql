query Processes($input: BpmnProcessesInput!) {
  processes(input: $input) {
    id
    name
  }
}
