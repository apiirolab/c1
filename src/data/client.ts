import * as QueryTypes from './generated/types';

export { default as gql } from 'graphql-tag';

export { QueryTypes };

export function purge<T>(params: any): T {
  for (let key of Object.getOwnPropertyNames(params)) {
    if (!params[key]) {
      delete params[key];
    }
    if (params[key] instanceof Object) {
      purge(params[key]);
    }
  }
  return params;
}
