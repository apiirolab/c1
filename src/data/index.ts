export const info = 'Data Info';
