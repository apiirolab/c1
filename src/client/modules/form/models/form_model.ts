// import * as shortid from 'shortid';
import * as validations from './validation';

import { IObservableArray } from 'mobx';
import { ISimpleType, types } from 'mobx-state-tree';

import { Yoga } from 'data/yoga';
import { ListValue } from './form_model';
import { FormValidation } from './form_validation_model';

export interface DataSet {
  id?: string;
  addRow(key: string): void;
  removeRow(key: string, index: number): void;
  removeRowData<T>(key: string, data: T): void;
  getList(name: string): ListValue[];
  getValue(name: string): any;
  parseValue(name: string, value: any): void;
  getError(name: string): any;
  getStringValue(name: string): any;
  setStringValue(name: string, value: any): void;
  isExpression(key: string): boolean;
}

export type FormControlProps = {
  formControl: Yoga.FormElement;
  owner: DataSet;
};

function formItemSort(a: Yoga.FormElement, b: Yoga.FormElement) {
  return a.row < b.row
    ? -1
    : a.row > b.row
      ? 1
      : a.column < b.column
        ? -1
        : a.column > b.column
          ? 1
          : 0;
}

function mstTypeFactory(
  desc: Yoga.DataDescriptor,
  lists: Array<{ name: string; items: any[] }>
): ISimpleType<any> {
  switch (desc.type) {
    case 'String':
      return types.string;
    case 'Int':
      return types.number;
    case 'Float':
      return types.number;
    case 'Boolean':
      return types.boolean;
    case 'Object':
      return FormModel.buildMst(desc.descriptors, lists);
    // return For
    case undefined:
      return types.string;
  }
  throw new Error('MST Type not supported: ' + desc.type);
}

export interface ListValue {
  [index: string]: string;
  text: string;
  value: string;
}

let time = Date.now();
let i = 0;

export class FormModel {
  static parseDefault(descriptor: Yoga.DataDescriptor) {
    switch (descriptor.type) {
      case 'Int':
        return parseInt(descriptor.defaultValue || '0', 10) as any;
      case 'Float':
        return parseFloat(descriptor.defaultValue || '0') as any;
      case 'Boolean':
        return (descriptor.defaultValue === 'true' || descriptor.defaultValue === 'True') as any;
      case 'Date':
        return new Date(descriptor.defaultValue);
    }
    return descriptor.defaultValue;
  }

  static buildMst(
    descriptors: Yoga.DataDescriptor[],
    lists?: Array<{ name: string; items: any[] }>
  ) {
    const descriptorMap: {
      [index: string]: Yoga.DataDescriptor;
    } = {};
    for (let d of descriptors) {
      descriptorMap[d.name] = d;
    }

    // prepare model and views
    const mstDefinition: { [index: string]: any } = {};
    const viewDefinition = (self: any) => {
      const view = {
        getDescriptor(key: string) {
          if (!descriptorMap[key]) {
            throw new Error('Descriptor does not exist: ' + key);
          }
          return descriptorMap[key];
        },
        isExpression(key: string) {
          return !!(key && this.getDescriptor(key).expression);
        },
        getValue(key: string) {
          if (self[key] == null) {
            return this.getDescriptor(key).defaultValue;
          }
          return self[key];
        },
        getList(key: string): ListValue[] {
          return self[key];
        },
        getStringValue(key: string) {
          // expressions are evaluated on the fly
          if (this.getDescriptor(key).expression) {
            // @ts-ignore
            return self[key];
          }
          return self[key + '_str'] || this.getDescriptor(key).defaultValue || '';
        },
        getError(key: string) {
          return self[key + '_error'];
        }
      };

      // add expressions views
      // expression are dynamic formulas evaluated from script

      for (let desc of descriptors) {
        // expressions do not need state tree entry they are evaluated automatically
        if (desc.expression) {
          (view as any).__defineGetter__(desc.name, function() {
            // tslint:disable-next-line:no-eval
            return eval(desc.expression);
          });
        }
      }

      if (lists) {
        for (let list of lists) {
          (view as any).__defineGetter__(list.name, function() {
            return list.items;
          });
        }
      }

      return view;
    };

    // add mst nodes for non expression fields
    // expressions do not need custom nodes and are handled by views
    for (let desc of descriptors) {
      // expressions do not need state tree entry they are evaluated automatically
      if (desc.isArray) {
        mstDefinition[desc.name] = types.array(mstTypeFactory(desc, lists));
      } else if (desc.type === 'Id') {
        mstDefinition[desc.name] = types.optional(types.identifier, () => (time + i++).toString()); // shortid.generate());
      } else if (!desc.expression) {
        mstDefinition[desc.name] = desc.defaultValue
          ? FormModel.parseDefault(desc)
          : types.maybe(mstTypeFactory(desc, lists));
        mstDefinition[desc.name + '_str'] = types.maybe(types.string);
        mstDefinition[desc.name + '_error'] = types.maybe(types.string);
      }
    }

    // add generic method

    // build tree
    const mst = types
      .model('Store', mstDefinition)
      .views(viewDefinition)
      .actions(self => ({
        addRow(key: string) {
          (self[key] as any[]).push({});
        },
        removeRow(key: string, index: number) {
          (self[key] as any[]).splice(index);
        },
        removeRowData<T>(key: string, data: T) {
          (self[key] as IObservableArray<T>).remove(data);
        },
        parseValue(key: string, value: any) {
          const descriptor = descriptorMap[key];
          switch (descriptor.type) {
            case 'Int':
              self[key] = parseInt(value || 0, 10) as any;
              break;
            case 'Float':
              self[key] = parseFloat(value || 0) as any;
              break;
            case 'Boolean':
              self[key] = (value === true || value === 'true' || value === 'True') as any;
              break;
            default:
              self[key] = value;
          }
        },
        setError(key: string, error: string) {
          (self as any)[key + '_error'] = error;
        }
      }))
      .actions(self => ({
        setStringValue(key: string, value: any) {
          self[key + '_str'] = value;

          const descriptor = descriptorMap[key];
          let error = '';
          switch (descriptor.type) {
            case 'Int':
              error = validations.intValidator(value);
              break;
            case 'Float':
              error = validations.floatValidator(value);
              break;
          }

          if (!error) {
            self.parseValue(key, value);
          }
          self.setError(key, error);
        }
      }));

    return mst;
  }

  static initStrings(data: any, dataArray: Array<{ name: string; value: any }>) {
    for (let item of dataArray) {
      data[item.name] = item.value;
      if (Array.isArray(item.value)) {
        data[item.name] = [];
        for (let m of item.value) {
          const n = {};
          data[item.name].push(n);
          FormModel.initStrings(n, m);
        }
      } else {
        data[item.name + '_str'] = item.value ? item.value.toString() : '';
      }
    }

    return data;
  }

  static buildMstModel(
    descriptors: Yoga.DataDescriptor[],
    dataArray: Array<{ name: string; value: any }>,
    lists?: Array<{ name: string; items: any[] }>
  ): DataSet {
    // data initialisation
    const data: any = {};
    FormModel.initStrings(data, dataArray);

    const mst = FormModel.buildMst(descriptors, lists);
    return mst.create(data);
  }

  id: string;
  name: string;
  description: string;
  elements: Yoga.FormElement[];
  validations: FormValidation[];

  constructor(form: Yoga.Form) {
    this.id = form.id;
    this.name = form.name;
    this.description = form.description;
    this.validations = form.validations;

    // add and sort elements
    this.elements = form.elements;
    this.elements.sort(formItemSort);
  }
}
