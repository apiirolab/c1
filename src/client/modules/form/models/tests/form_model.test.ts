import { autorun } from 'mobx';

import { create } from 'client/modules/form/views/tests/form_query_data';
import { QueryTypes } from 'data/client';
import { FormModel } from '../form_model';

it('creates a new model', () => {
  let model = new FormModel({
    id: '1',
    description: 'Desc',
    elements: [],
    name: 'Name',
    validations: []
  });

  expect(model.id).toEqual('1');
  expect(model.description).toEqual('Desc');
  expect(model.elements).toEqual([]);
  expect(model.name).toEqual('Name');
  expect(model.validations).toEqual([]);
});

it('sorts form fields when form is created', () => {
  let model = new FormModel(
    create.form({
      elements: [
        create.formElement({ id: '6', row: 3, column: 2, width: 6 }),
        create.formElement({ id: '5', row: 3, column: 2, width: 6 }),
        create.formElement({ id: '2', row: 0, column: 8, width: 6 }),
        create.formElement({ id: '3', row: 2, column: 2, width: 6 }),
        create.formElement({ id: '4', row: 2, column: 8, width: 6 }),
        create.formElement({ id: '1', row: 0, column: 0, width: 6 })
      ]
    })
  );
  expect(model.elements[0].id).toEqual('1');
  expect(model.elements[1].id).toEqual('2');
  expect(model.elements[2].id).toEqual('3');
  expect(model.elements[3].id).toEqual('4');
  expect(model.elements[4].id).toEqual('6');
});

it('builds MST', () => {
  const descriptors = [
    create.descriptor({ name: 'age', type: QueryTypes.DataType.Int, defaultValue: '7' }),
    create.descriptor({ name: 'height', type: QueryTypes.DataType.Int, defaultValue: '5' }),
    create.descriptor({ id: 'p', name: 'parent', type: QueryTypes.DataType.Object }),
    create.descriptor({
      name: 'weight',
      parentDescriptor: 'p',
      type: QueryTypes.DataType.Int,
      defaultValue: '60'
    }),
    create.descriptor({
      name: 'taller',
      type: QueryTypes.DataType.Int,
      expression: `this['height'] + 10`
    }),
    create.descriptor({
      id: '1',
      name: 'countries',
      type: QueryTypes.DataType.Object,
      isArray: true
    }),
    create.descriptor({ name: 'name', type: QueryTypes.DataType.String, parentDescriptor: '1' }),
    create.descriptor({ name: 'capital', type: QueryTypes.DataType.String, parentDescriptor: '1' })
  ];

  const instance = FormModel.buildMstModel(descriptors, {
    height: 6,
    parent: { weight: 10 },
    countries: [{ name: 'Slovakia', capital: 'Bratislava' }]
  });

  expect(instance.getValue('height')).toEqual(6);
  expect(instance.getValue('taller')).toEqual(16);

  let finalHeight = 0;
  autorun(() => {
    finalHeight = instance.getValue('taller');
  });

  // check computed fields
  instance.parseValue('height', 10);
  expect(finalHeight).toEqual(20);

  // check js version
  expect(instance.toJS()).toMatchSnapshot();
});
