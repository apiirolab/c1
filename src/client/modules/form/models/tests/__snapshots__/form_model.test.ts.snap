// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds MST 1`] = `
Object {
  "age": 7,
  "countries": Array [
    Object {
      "capital": "Bratislava",
      "defaultValue": Object {
        "capital": undefined,
        "name": undefined,
        "strings": Object {
          "capital": "",
          "name": "",
        },
      },
      "dirty": false,
      "getDescriptor": [Function],
      "isExpression": [Function],
      "name": "Slovakia",
    },
  ],
  "defaultValue": Object {
    "age": "7",
    "countries": Array [],
    "countries_default": Object {
      "capital": undefined,
      "name": undefined,
      "strings": Object {
        "capital": "",
        "name": "",
      },
    },
    "height": "5",
    "parent": Object {
      "strings": Object {
        "weight": "60",
      },
      "weight": "60",
    },
    "strings": Object {
      "age": "7",
      "height": "5",
    },
  },
  "dirty": true,
  "getDescriptor": [Function],
  "height": 10,
  "history": Object {
    "history": Array [
      Object {
        "inversePatches": Array [
          Object {
            "op": "replace",
            "path": "/height",
            "value": 6,
          },
          Object {
            "op": "replace",
            "path": "/dirty",
            "value": false,
          },
        ],
        "patches": Array [
          Object {
            "op": "replace",
            "path": "/height",
            "value": 10,
          },
          Object {
            "op": "replace",
            "path": "/dirty",
            "value": true,
          },
        ],
      },
    ],
    "undoIdx": 1,
  },
  "isExpression": [Function],
  "parent": Object {
    "defaultValue": Object {
      "strings": Object {
        "weight": "60",
      },
      "weight": "60",
    },
    "dirty": false,
    "getDescriptor": [Function],
    "isExpression": [Function],
    "weight": 10,
  },
}
`;
